### PR TITLE
feat!: Update autoscaling group `tags` -> `tag` to support v4 of AWS provider

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- Update autoscaling group `tags` -> `tag` to support v4 of AWS provider which is now a tag block as opposed to a standard attribute
- `var.propagate_name` has been removed
- `var.tags_as_map` has been moved into `var.tags` and now `var.tags` just takes the standard `map(string)` map of key value pairs for tags. The tags provided are propagated to instances launched through the dynamic tag block
- fix example for predictive scaling policy which caught some issues in the nested map lookup logic that has been corrected
- `var.tags` are now automatically merged with `var.tag_specifications`

## Motivation and Context
- Avoids showing users a deprecation statement for `tags` on autoscaling group
- Simplify tag usage and removes constant tag updates in each plan/apply

## Breaking Changes
- Yes
	- `var.propagate_name` has been removed
	- `var.tags_as_map` has been moved into `var.tags` and now `var.tags` just takes the standard `map(string)` map of key value pairs for tags. The tags provided are propagated to instances launched through the dynamic tag block
	
	Users will need to ensure that the value they are passing to `var.tags` is the standard map of key=value pairs (`map(string)`)

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -446,7 +446,14 @@ resource "aws_autoscaling_group" "this" {
     delete = var.delete_timeout
   }
 
-  tags = local.tags
+  dynamic "tag" {
+    for_each = { for tag in local.tags : "${tag.key}-${tag.value}" => tag }
```
[the remark this answers is not in the document]
Comment (antonbabenko) [reply main.tf:448]: Right, I wonder what will be the result if a user provides keys in several ways like this: `{Project = "foo"}` and `{Project = "bar"}`.

Without deduplication, there will be 2 keys in `for_each` created - `Project-foo` and `Project-bar`. Will it lead to constant changes after the creation?

[the remark this answers is not in the document]
Comment (antonbabenko) [reply main.tf:448]: Let's do it to reduce the number of potential issues related to constant changes.

File: main.tf [hunk 2]
```diff
@@ -542,28 +532,28 @@ resource "aws_autoscaling_policy" "this" {
       scheduling_buffer_time       = lookup(predictive_scaling_configuration.value, "scheduling_buffer_time", null)
 
       dynamic "metric_specification" {
-        for_each = lookup(predictive_scaling_configuration.value, "metric_specification", [])
+        for_each = can(predictive_scaling_configuration.value.metric_specification.target_value) ? [predictive_scaling_configuration.value.metric_specification] : []
```
Comment (bryantbiggs) [inline main.tf:535]: with the bug fix on line 560, I had to make some adjustments to get this to work as intended and without errors

File: variables.tf [hunk 3]
```diff
@@ -212,23 +212,11 @@ variable "delete_timeout" {
...
 
 variable "tags" {
-  description = "A list of tag blocks. Each element should have keys named key, value, and propagate_at_launch"
-  type        = list(map(string))
-  default     = []
-}
-
-variable "tags_as_map" {
-  description = "A map of tags and values in the same format as other resources accept. This will be converted into the non-standard format that the aws_autoscaling_group requires."
+  description = "A map of tags to assign to resources"
```
Comment (bryantbiggs) [inline variables.tf:215]: brings everything back inline with other modules where users just provide `var.tags`

**Comment by bryantbiggs** [comment]
ok should be ready for review now if you get some time @antonbabenko

**Comment by antonbabenko** [comment]
@bryantbiggs I am pretty much off with corona since Wednesday. Only obvious PRs can be processed by me before Monday 🤯

**Comment by bryantbiggs** [comment]
@antonbabenko no rush at all - sorry to hear, hope you feel better soon!

File: examples/complete/main.tf [hunk 4]
```diff
@@ -557,7 +531,11 @@ module "complete_lt" {
     },
     predictive-scaling = {
       policy_type = "PredictiveScaling"
-      predictive_scaling_config = {
```
[the remark this answers is not in the document]
Comment (antonbabenko) [reply examples/complete/main.tf:560]: Probably nobody was using it :)

**Review by antonbabenko — APPROVED** [review]
Looks good. To make this release a "breaking change" I need to include "BREAKING CHANGE" into the commit message (if I remember correctly). Fingers crossed!

**Comment by antonbabenko** [comment]
This PR is included in version 5.0.0 :tada:

**Pull request merged**